Tiltfile

```diff
@@ -0,0 +1,190 @@
+# -*- mode: Python -*-
+
+k8s_yaml([
+'config/crd/bases/velero.io_backups.yaml',
+'config/crd/bases/velero.io_backupstoragelocations.yaml', 
+'config/crd/bases/velero.io_deletebackuprequests.yaml',   
+'config/crd/bases/velero.io_downloadrequests.yaml',   
+'config/crd/bases/velero.io_podvolumebackups.yaml',  
+'config/crd/bases/velero.io_podvolumerestores.yaml',  
+'config/crd/bases/velero.io_resticrepositories.yaml',
+'config/crd/bases/velero.io_restores.yaml',    
+'config/crd/bases/velero.io_schedules.yaml', 
+'config/crd/bases/velero.io_serverstatusrequests.yaml',  
+'config/crd/bases/velero.io_volumesnapshotlocations.yaml', 
+])
+
+# default values
+settings = {
+   "default_registry": "",
+   "enable_restic": False,
+   "create_backup_locations": False, 
+   "setup-minio": False, 
+}
+
+# global settings
+settings.update(read_json(
+    "tilt-resources/tilt-settings.json",
+    default = {},
+))
+
+k8s_yaml(kustomize('tilt-resources'))
+k8s_yaml('tilt-resources/deployment.yaml')
+if settings.get("enable_restic"):
+    k8s_yaml('tilt-resources/restic.yaml')
+if settings.get("create_backup_locations"):
+    k8s_yaml('tilt-resources/backupstoragelocations.yaml')
+if settings.get("setup-minio"):
+    k8s_yaml('examples/minio/00-minio-deployment.yaml', allow_duplicates=True)
+
+# By default, Tilt automatically allows Minikube, Docker for Desktop, Microk8s, Red Hat CodeReady Containers, Kind, K3D, and Krucible.
+allow_k8s_contexts(settings.get("allowed_contexts"))
+default_registry(settings.get("default_registry"))
+
+tilt_helper_dockerfile_header = """
+# Tilt image
+FROM golang:1.15.3 as tilt-helper
+# Support live reloading with Tilt
+RUN wget --output-document /restart.sh --quiet https://raw.githubusercontent.com/windmilleng/rerun-process-wrapper/master/restart.sh  && \
+    wget --output-document /start.sh --quiet https://raw.githubusercontent.com/windmilleng/rerun-process-wrapper/master/start.sh && \
+    chmod +x /start.sh && chmod +x /restart.sh
+"""
+
+additional_docker_helper_commands = """
+RUN wget -qO- https://dl.k8s.io/v1.19.2/kubernetes-client-linux-amd64.tar.gz | tar xvz
+RUN wget -qO- https://get.docker.com | sh
+"""
+
+additional_docker_build_commands = """
+COPY --from=tilt-helper /usr/bin/docker /usr/bin/docker
+COPY --from=tilt-helper /go/kubernetes/client/bin/kubectl /usr/bin/kubectl
+"""
+
+##############################
+# Setup Velero
+##############################
+
+# Set up a local_resource build of the Velero binary. The binary is written to .tiltbuild/velero.
+local_resource(
+    "velero_manager",
+    cmd = "cd " + "." + ';mkdir -p .tiltbuild;CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -ldflags \'-extldflags "-static"\' -o .tiltbuild/velero ' + './cmd/velero',
```

Comment:
Any way we can get this to use `make local`? I'm not sure I like that it's using a different build command than what we normally use, as that could introduce discrepancies.

Reply:
I'm sure we can, but that would defeat the purpose. In that case, we could use `make local` and not bother with Tilt. The goal here is to make the leanest possible build. Let's think about what discrepancy could be introduced that would make this not feasible. This tool is rebuilding the binary on each change. But not even the entire binary, only the parts specified in the `deps` setting. The assumption is you don't need to see anything else but the code you are working on.

Reply:
`not even the entire binary` - sorry I was up way too early today :) 

Of course if rebuilds the entire binary, but it only looks for changes in the files/directories that we specify, changes in other parts don't get included in the rebuild.

Reply:
I'm not sure I follow how using `make local` here defeats the purpose of Tilt, since it's just using our Makefile arguments outside of a container. Though I do see that `make local` would need to have different arguments passed if you're working on macOS or Windows, though, since we'd always be making a Linux build.

Here's where I see possible discrepancies - if we ever need to change `go build` flags, we'd have to update them in the Tiltfile and the Makefile with this setup. That's something easy to miss unless we set up automation for it.

Since this line is about `go build`, I'd think we could use the `deps` within our `make local` target too, couldn't we? Or is it about what files Tilt watches to trigger the command run?

Reply:
Yes, "on save" of files/dirs marked to be watched Tilt gets triggered to rebuild automatically.

Reply:
Let's touch base on this, I don't think I'm understanding what you mean.

When I said using `make container` would defeat the purpose, that's because it takes over 4 min to run:
![image](https://user-images.githubusercontent.com/16508/100766785-e3415300-33ad-11eb-8318-65233c19f802.png)

Vs. a few seconds with this Tilt setup:
![image](https://user-images.githubusercontent.com/16508/100766930-0b30b680-33ae-11eb-8dd2-9ae4e0263cf8.png)

At this point, the fact that the cmd gets triggered automatically on file changes is hardly a win.

Again, I might not be understanding what you are suggesting so, let's talk about it.

Reply:
`make container` isn't `make local` :) The `make local` target should only produce a binary, not a container.

This is how `make local` runs on my machine:

```shell
% time make local
GOOS=darwin \
	GOARCH=amd64 \
	VERSION=main \
	PKG=github.com/vmware-tanzu/velero \
	BIN=velero \
	GIT_SHA=aa4730970037ce293a0df845381d18d172cdc6d3 \
	GIT_TREE_STATE=dirty \
	OUTPUT_DIR=$(pwd)/_output/bin/darwin/amd64 \
	./hack/build.sh
make local  51.18s user 11.23s system 467% cpu 13.339 total
```

If we just use `hack/build.sh`, (which the `make local` target invokes) with minimal arguments, then we get a much lower time:

```
% time PKG=github.com/vmware-tanzu/velero BIN=velero GOOS=linux GOARCH=amd64 VERSION=main GIT_SHA=aa4730970037ce29 GIT_TREE_STATE=dirty OUTPUT_DIR=tiltout ./hack/build.sh
PKG=github.com/vmware-tanzu/velero BIN=velero GOOS=linux GOARCH=amd64 VERSION  
2.55s user 2.07s system 788% cpu 0.586 total
```

`make local` is slower because it's looking up the git SHA each time, as well as doing some make target dependencies like ensuring the build directories exist. For these purposes, we might be able to skip the git SHA and just say that the tree is dirty so we know we're not building a "clean" binary. All `build.sh` is really doing is making sure some variables are set then invoking the `go build` command.

Reply:
My brain had read *make local* as *make container*, no wonder what you said wasn't making sense to me. Sorry about that. I see your point clearly now, let me see if I can make that work.